When the deployment of a process definition failed, clients can still start new process instances of the actually non-existing process definition

**Describe the bug**

When the command to deploy processes failed with, for example, `EXCEEDED_BATCH_RECORD_SIZE`, any processes - that have been cached during the deployment process - are still kept in cache afterward. That way, clients can start process instances of these processes. Once a leader change happened (or the broker restarted), the Stream Process fails to replay these started process instances:

```
java.lang.RuntimeException: java.lang.IllegalStateException: Expected to find a process deployed with key '2251799813685405' but not found.
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$replayNextEvent$4(ReplayStateMachine.java:169) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) [zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) [zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) [zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) [zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:198) [zeebe-scheduler-8.2.4.jar:8.2.4]
Caused by: java.lang.IllegalStateException: Expected to find a process deployed with key '2251799813685405' but not found.
	at io.camunda.zeebe.engine.state.deployment.DbProcessState.getFlowElement(DbProcessState.java:288) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.state.appliers.ProcessInstanceElementActivatingApplier.createEventScope(ProcessInstanceElementActivatingApplier.java:261) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.state.appliers.ProcessInstanceElementActivatingApplier.applyState(ProcessInstanceElementActivatingApplier.java:51) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.state.appliers.ProcessInstanceElementActivatingApplier.applyState(ProcessInstanceElementActivatingApplier.java:32) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.state.appliers.EventAppliers.applyState(EventAppliers.java:302) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.Engine.replay(Engine.java:110) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$replayEvent$7(ReplayStateMachine.java:230) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at java.util.Optional.ifPresent(Unknown Source) ~[?:?]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.replayEvent(ReplayStateMachine.java:230) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at java.util.Iterator.forEachRemaining(Unknown Source) ~[?:?]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$tryToReplayBatch$5(ReplayStateMachine.java:208) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:84) ~[zeebe-db-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.tryToReplayBatch(ReplayStateMachine.java:206) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$replayNextEvent$3(ReplayStateMachine.java:165) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:28) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.retry.RecoverableRetryStrategy.run(RecoverableRetryStrategy.java:48) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:92) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	... 5 more
```

Basically, when processing a deployment command the following happens

1. `DeploymentCreateProcessor#processRecord()` transforms the given resources accordingly
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java#L129-L135
2. If succeeded, the engine will create timers for any timer start event
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java#L137-L142
3. When executing the `#createTimerIfTimerStartEvent()` it will try to get the process definition by key:
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java#L193-L203
4. Getting the process definition by key will eventually put the process definition into a map `processesByProcessIdAndVersion` for caching purposes:
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java#L201-L211
5. Once `#createTimerIfTimerStartEvent()` finishes, the engine will write a follow-up event to indicate the deployment has been created
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java#L146
6. Writing this follow-up event might fail with `EXCEEDED_BATCH_RECORD_SIZE`

As a consequence, the RocksDB transaction gets rolled back, meaning, the process definitions are not stored in RocksDB. But the not successful deployed process definitions are still present in the map `processesByProcessIdAndVersion`. That way, when a client starts a process by `bpmnProcessId`, the engine will get the actually non-existing process definition from the map `processesByProcessIdAndVersion` and start a process instance eventually. And as long as the leader keeps the same and Zeebe does not restart, process instances of the non-existing process definition can get started.

Additionally, the `ProcessVersionManager#versionCache` caches the latest version of a process definition. The same applies here, when the deployment fails, the cached latest version has been already increased and stays present in the cache. Basically, since `processesByProcessIdAndVersion` contains a non-existing process definition (for example, with version `12`) and the `versionCache` points to the non-existing version `12` of that process definition, the engine will return the process definition from `processesByProcessIdAndVersion`

https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java#L214-L231

Before `versionCache` have been introduced, the engine would get the latest version from RocksDB which would not be `12` for sure. That way, the engine ensured that it always returns a process definition that actually exists.

**To Reproduce**
1. Deploy many processes at once (containing a process with `bpmnProcessId=foo` )
2. Ensure that the deployment fails with `EXCEEDED_BATCH_RECORD_SIZE`
3. Start a process instance by `bpmnProcessId=foo`  => The process instance gets started accordingly.
4. Stop the broker
5. Start the broker

**Expected behavior**
* Whenever a deployment fails, the client cannot start any process instance of any of those process definitions.
* To be more concrete, any process definitions that get cached when processing the deployment command must be removed when the deployment failed.
* Or as an alternative, don't cache any process definitions during deployment.
* The same applies to `ProcessVersionManager#versionCache`

**What is the impact of the issue?**
* Any follower cannot replay any events of process instances belonging to a non-existing process definition. That way, the stream processor will stop replaying anything. The follower won't create any new snapshots, and so on.
* If there is a failover, the new leader will fail to replay the events when recovering from the last snapshot. It will again stop the stream processor. No snapshots are taken anymore. The back pressure goes up to 100%. The partition gets unhealthy.
* It may result in process definitions that are available on partition >= 2 but not on partition 1. For example:
  1. Deploy the first version of a process `foo` that succeeds
  2. Deploy a new version of the process `foo` that fails with `EXCEEDED_BATCH_RECORD_SIZE` (try doing multiple times)
  3. Then redeploy the first version once again (and let it succeed)
  4. => It will result in a process definition with version `x` (`x > 1`) which gets distributed to other partitions but on partition 1 this process definition won't get persisted in RocksDB.
  5. So on partition > 1, there are running process instances referencing process definition with version `x`. But these process instances cannot be opened in Operate, because the process definition was never exported and imported.
* There is no way to recover from that. Only purging the clusters resolves the issue.

**Environment:**
- Zeebe Version: 8.2.4

---

related to #5723 
related to [SUPPORT-17483](https://jira.camunda.com/browse/SUPPORT-17483)
related to #12873

--- a/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java
+++ b/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java
@@ -96,6 +96,9 @@ public final class DeploymentCreateProcessor implements TypedRecordProcessor<Dep
     try {
       createTimerIfTimerStartEvent(command);
     } catch (final RuntimeException e) {
+      // Make sure the cache does not contain any leftovers from this run (by hard resetting)
+      processState.clearCache();
+
       final String reason = String.format(COULD_NOT_CREATE_TIMER_MESSAGE, e.getMessage());
       responseWriter.writeRejectionOnCommand(command, RejectionType.PROCESSING_ERROR, reason);
       rejectionWriter.appendRejection(command, RejectionType.PROCESSING_ERROR, reason);
@@ -116,6 +119,9 @@ public final class DeploymentCreateProcessor implements TypedRecordProcessor<Dep
   @Override
   public ProcessingError tryHandleError(
       final TypedRecord<DeploymentRecord> command, final Throwable error) {
+    // Make sure the cache does not contain any leftovers from this run (by hard resetting)
+    processState.clearCache();
+
     if (error instanceof ResourceTransformationFailedException exception) {
       rejectionWriter.appendRejection(
           command, RejectionType.INVALID_ARGUMENT, exception.getMessage());

--- a/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java
+++ b/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java
@@ -302,6 +302,13 @@ public final class DbProcessState implements MutableProcessState {
     return element;
   }
 
+  @Override
+  public void clearCache() {
+    processesByKey.clear();
+    processesByProcessIdAndVersion.clear();
+    versionManager.clear();
+  }
+
   private DeployedProcess lookupProcessByIdAndPersistedVersion(final long latestVersion) {
     processVersion.wrapLong(latestVersion);
 

--- a/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/ProcessVersionManager.java
+++ b/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/ProcessVersionManager.java
@@ -69,4 +69,8 @@ public final class ProcessVersionManager {
     }
     return currentValue;
   }
+
+  public void clear() {
+    versionCache.clear();
+  }
 }

--- a/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/ProcessState.java
+++ b/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/ProcessState.java
@@ -30,4 +30,7 @@ public interface ProcessState {
 
   <T extends ExecutableFlowElement> T getFlowElement(
       long processDefinitionKey, DirectBuffer elementId, Class<T> elementType);
+
+  /** TODO: Remove the cache entirely from the immutable state */
+  void clearCache();
 }

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
@@ -302,6 +302,7 @@ public class DeploymentRejectionTest {
     final BpmnModelInstance invalidProcess =
         Bpmn.createExecutableProcess("too_large_process")
             .startEvent()
+            // In order to cause BATCH SIZE EXCEEDING we add a big comment
             .documentation("x".repeat((int) ByteValue.ofMegabytes(3)))
             .done();
     final BpmnModelInstance validProcess =

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
@@ -297,7 +297,7 @@ public class DeploymentRejectionTest {
 
   /** Regression test against https://github.com/camunda/zeebe/issues/13254 */
   @Test
-  public void shouldNotCacheProcessesWhenDeploymentRejected() {
+  public void shouldNotBeAbleToCreateInstanceWhenDeploymentIsRejected() {
     // given
     final BpmnModelInstance invalidProcess =
         Bpmn.createExecutableProcess("too_large_process")

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
@@ -23,7 +23,6 @@ import io.camunda.zeebe.protocol.record.intent.DeploymentDistributionIntent;
 import io.camunda.zeebe.protocol.record.intent.DeploymentIntent;
 import io.camunda.zeebe.protocol.record.intent.ProcessIntent;
 import io.camunda.zeebe.protocol.record.value.DeploymentRecordValue;
-import io.camunda.zeebe.test.util.junit.RegressionTest;
 import io.camunda.zeebe.test.util.record.RecordingExporter;
 import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
 import io.camunda.zeebe.util.ByteValue;
@@ -296,7 +295,7 @@ public class DeploymentRejectionTest {
             tuple(DeploymentDistributionIntent.DISTRIBUTING, RecordType.EVENT));
   }
 
-  @RegressionTest("https://github.com/camunda/zeebe/issues/13254")
+  @Test // Regression of https://github.com/camunda/zeebe/issues/13254
   public void shouldNotBeAbleToCreateInstanceWhenDeploymentIsRejected() {
     // given
     final BpmnModelInstance invalidProcess =

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
@@ -294,4 +294,37 @@ public class DeploymentRejectionTest {
             tuple(DeploymentIntent.CREATED, RecordType.EVENT),
             tuple(DeploymentDistributionIntent.DISTRIBUTING, RecordType.EVENT));
   }
+
+  /** Regression test against https://github.com/camunda/zeebe/issues/13254 */
+  @Test
+  public void shouldNotCacheProcessesWhenDeploymentRejected() {
+    // given
+    final BpmnModelInstance invalidProcess =
+        Bpmn.createExecutableProcess("too_large_process")
+            .startEvent()
+            .documentation("x".repeat((int) ByteValue.ofMegabytes(3)))
+            .done();
+    final BpmnModelInstance validProcess =
+        Bpmn.createExecutableProcess("valid_process").startEvent().task().endEvent().done();
+
+    // when
+    ENGINE
+        .deployment()
+        .withXmlResource(invalidProcess)
+        .withXmlResource(validProcess)
+        .expectRejection()
+        .deploy();
+
+    // then
+    assertThat(
+            RecordingExporter.records()
+                .limit(r -> r.getRecordType() == RecordType.COMMAND_REJECTION)
+                .collect(Collectors.toList()))
+        .extracting(Record::getIntent, Record::getRecordType)
+        .doesNotContain(
+            tuple(ProcessIntent.CREATED, RecordType.EVENT),
+            tuple(DeploymentIntent.CREATED, RecordType.EVENT));
+
+    ENGINE.processInstance().ofBpmnProcessId("valid_process").expectRejection().create();
+  }
 }

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java
@@ -23,6 +23,7 @@ import io.camunda.zeebe.protocol.record.intent.DeploymentDistributionIntent;
 import io.camunda.zeebe.protocol.record.intent.DeploymentIntent;
 import io.camunda.zeebe.protocol.record.intent.ProcessIntent;
 import io.camunda.zeebe.protocol.record.value.DeploymentRecordValue;
+import io.camunda.zeebe.test.util.junit.RegressionTest;
 import io.camunda.zeebe.test.util.record.RecordingExporter;
 import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
 import io.camunda.zeebe.util.ByteValue;
@@ -295,8 +296,7 @@ public class DeploymentRejectionTest {
             tuple(DeploymentDistributionIntent.DISTRIBUTING, RecordType.EVENT));
   }
 
-  /** Regression test against https://github.com/camunda/zeebe/issues/13254 */
-  @Test
+  @RegressionTest("https://github.com/camunda/zeebe/issues/13254")
   public void shouldNotBeAbleToCreateInstanceWhenDeploymentIsRejected() {
     // given
     final BpmnModelInstance invalidProcess =

